odl/space/cu_ntuples.py

```diff
@@ -807,6 +820,122 @@ def fun(x, out=None):
 sqrt = _make_unary_fun('sqrt')
 
 
+def _weighted(weight, attr, exponent, dist_using_inner=False):
+    if np.isscalar(weight):
+        weighting = CudaFnConstWeighting(
+            weight, exponent)
+    elif isinstance(weight, CudaFn.Vector):
+        weighting = CudaFnVectorWeighting(
+            weight, exponent=exponent, dist_using_inner=dist_using_inner,
+            copy_to_gpu=True)
+    else:
+        weight_ = np.asarray(weight)
+        if weight_.dtype == object:
+            raise ValueError('bad weight {}'.format(weight))
+        if weight_.ndim == 1:
+            weighting = CudaFnVectorWeighting(
+                weight_, exponent, dist_using_inner=dist_using_inner)
+        elif weight_.ndim == 2:
+            raise NotImplementedError('matrix weighting not implemented '
+                                      'for CUDA spaces.')
+#            weighting = CudaFnMatrixWeighting(
+#                weight_, exponent, dist_using_inner=dist_using_inner)
+        else:
+            raise ValueError('array-like weight must have 1 or 2 dimensions, '
+                             'but {} has {} dimensions.'
+                             ''.format(weight, weight_.ndim))
+    return getattr(weighting, attr)
```

Comment:
Well, the point of the free functions below was to get the weighted inner product, norm and dist without having to to through the creation of the weighting class explicitly. Pure convenience. Code-wise this is just DRY since for each of the three free functions, the procedure is the same - create the class and fish out the corresponding attribute.